src/Core/Services/Implementations/AzureQueueReferenceEventService.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System.Threading.Tasks;
+using Azure.Storage.Queues;
+using Bit.Core.Models;
+using Newtonsoft.Json;
+
+namespace Bit.Core.Services
+{
+    public class AzureQueueReferenceEventService : IReferenceEventService
+    {
+        private const string QueueName = "reference-events";
+
+        private readonly QueueClient _queueClient;
+        private readonly GlobalSettings _globalSettings;
+
+        public AzureQueueReferenceEventService (
+            GlobalSettings globalSettings)
+        {
+            _queueClient = new QueueClient(globalSettings.Storage.ConnectionString, QueueName);
+            _globalSettings = globalSettings;
+        }
+
+        public async Task RaiseEventAsync(IReferenceable reference, string eventType,
+            object additionalInfo = null)
+        {
+            await SendMessageAsync(reference, eventType, additionalInfo);
+        }
+
+        private async Task SendMessageAsync(IReferenceable reference, string eventType,
+            object additionalInfo)
+        {
+            if (_globalSettings.SelfHosted)
+            {
+                return;
+            }
```

Comment:
Should never happen right?

Reply:
I have a tendency to lean towards overprotective "safe" coding when it comes to DI and encapsulation (never assume the caller has done its own diligence), but in theory no, this should never happen. Did you want me to remove this check?

Reply:
you can leave it